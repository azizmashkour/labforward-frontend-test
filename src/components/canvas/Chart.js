import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ChartJs from 'chart.js';

export default class Chart extends Component {
  constructor(props) {
    super(props);
    this.chartRef = React.createRef();
  }

  componentDidMount() {
    const {
      backgroundColor,
      color,
      data,
      fill,
      label,
      steppedLine,
      stepSize,
    } = this.props;

    const labels = Array.from(Array(data.length), (x, i) => i + 1);

    this.chartRef.current.height = 100;

    this.myChart = new ChartJs(this.chartRef.current, {
      type: 'line',
      data: {
        labels,
        datasets: [{
          backgroundColor: backgroundColor || color,
          data,
          fill,
          label,
          borderColor: color,
          lineTension: 0,
          steppedLine,
        }],
      },
      options: {
        elements: {
          point: {
            radius: 0,
          },
        },
        scales: {
          yAxes: [{
            ticks: {
              beginAtZero: true,
              stepSize: stepSize || 4,
            },
          }],
        },
      },
    });
  }

  componentDidUpdate() {
    const { data } = this.props;
    const labels = Array.from(Array(data.length), (x, i) => i + 1);
    this.myChart.labels = labels;
    this.myChart.data.datasets[0].data = data;
    this.myChart.update();
  }

  render() {
    return (
      <canvas ref={this.chartRef} />
    );
  }
}

Chart.propTypes = {
  backgroundColor: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  data: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
  fill: PropTypes.bool.isRequired,
  label: PropTypes.string.isRequired,
  steppedLine: PropTypes.bool.isRequired,
  stepSize: PropTypes.number.isRequired,
};
