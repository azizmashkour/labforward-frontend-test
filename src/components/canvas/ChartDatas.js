import React from 'react';
import PropTypes from 'prop-types';
import Chart from './Chart';

const ChartDatas = ({ data, threshold, t }) => {
  // handling peak detection function
  const handlePeakDetection = (data, threshold) => {
    const parsedDatas = [];
    let last = 0;
    // looping around all our datas
    for (let i = 1; i < data.length; i++) {
      // here we check if the current data minus the previous one is >= threshold
      // then the last element will be the 0 else 1
      // so our parsedDatas array will contain the last element
      // finally we just return our parsedDatas array.
      if (Math.abs(data[i] - data[i - 1]) >= threshold) {
        last = last === 0 ? 1 : 0;
      }
      parsedDatas.push(last);
    }
    return parsedDatas;
  };

  const datas = handlePeakDetection(data, threshold);

  return (
    <>
      <Chart
        backgroundColor="rgb(255, 99, 132, 0.5)"
        color="rgb(255, 99, 132)"
        data={data}
        fill={false}
        label="Actual Data"
        steppedLine={false}
        stepSize={2}
      />
      <Chart
        backgroundColor="rgba(54, 162, 235, 0.5)"
        color="rgb(54, 162, 235)"
        data={datas}
        fill={true}
        label="Areas of interest"
        steppedLine={false}
        stepSize={0.25}
      />
    </>
  );
};

ChartDatas.propTypes = {
  data: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
  threshold: PropTypes.number.isRequired,
};

export default ChartDatas;
