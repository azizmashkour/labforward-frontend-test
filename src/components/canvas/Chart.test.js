import React from 'react';
import { shallow } from 'enzyme';
import Chart from './Chart';
import ChartDatas from './ChartDatas';

describe('<Chart />', () => {
  const props = {
    data: [1, 0, 3, 9, 0],
    backgroundColor: 'rgb(0, 2, 255, .5)',
    threshold: 5,
  }
  it('renders the Charts component', () => {
    const component = shallow(<ChartDatas {...props}/>);
    expect(component.find(Chart).length).toBe(2);
    expect(component).toMatchSnapshot();
  });
});
