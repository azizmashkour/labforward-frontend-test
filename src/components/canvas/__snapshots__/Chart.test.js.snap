// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> renders the Chart component 1`] = `ShallowWrapper {}`;

exports[`<Chart /> renders the Charts component 1`] = `ShallowWrapper {}`;

exports[`renders the Chart component 1`] = `ShallowWrapper {}`;
